### ISSUE TITLE
Automatically remove hidden waypoints/caches from individual route

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#12387]
Lately I used the individual route feature to plan the best routing for a long multi cache with given wapoints.
While out following the route I stumbled over the following problem:
- My map filter is set up to hide visited waypoints
- When reaching a waypoint I complete the task, open the waypoint popup to mark it as visited
- This will automatically hide it from the map

Now I have no easy possibility to long tap it to remove it from my route and thus trigger setting the next (then first) routing point as target.
I think we can automate, that if the user selected to hide visited waypoints and sets a waypoint which is part of a route to visited, it shall be removed from that route.

@moving-bits What do you think?

BTW:This could also apply to the found status of caches (if found caches are set to be hidden and user sets a caches as found). In general to avoid routes to end up on an empty spot of the map.

**Comment by moving-bits** [comment]
Need to think about some automatism. Currently you have the following options:
- either reverse your workflow: first remove the visited waypoint from your route, then mark it as visited
- or open "route sort window" using the quick access button, and remove the top-most entry

**Comment by Lineflyer** [comment]
Yeah, I know. But most of the time I only remember that a visited waypoint is part of my route (and not just the route coincidently passing it) after setting to visited and seeing this strange route which ends in nowhere.

**Comment by moving-bits** [comment]
@eddiemuc 
When I'm in a `WaypointPopupFragment` or `CachePopupFragment`, how can I find out, whether "hide visited waypoints" or "hide found caches (including "offline found") is currently active? 
- "hide visited waypoints" does not seem to be the global preference (`Settings.getHideVisitedWaypoints()`).
- and how can I address a method from the calling activity (either `NewMap` or `CGeoMap`)?

**Comment by eddiemuc** [comment]
> @eddiemuc When I'm in a `WaypointPopupFragment` or `CachePopupFragment`, how can I find out, whether "hide visited waypoints" or "hide found caches (including "offline found") is currently active?
> 
> * "hide visited waypoints" does not seem to be the global preference (`Settings.getHideVisitedWaypoints()`).
> * and how can I address a method from the calling activity (either `NewMap` or `CGeoMap`)?

Hi @moving-bits , 

To find out about the "hide found caches" (or any "quickfilter" setting used in the map popup), you have to get the current filter for the context you're interested in. This you do via `GeocacheFilterContext` (which is an enum). There are three contexts defined currently: `LIVE`, `OFFLINE` and `TRANSIENT` - but for the third one there's no permanent storage. The current filter for a context (as a `GeocacheFilter` object) is retrieved via e.g. `GeocacheFilterContext.LIVE.get()`

Once you have a `GeocacheFilter` object in your hand, then all quickfilter settings are retrieved from it via method `getQuickFilter()`. Available quickfilter keys are defined in enum `GeocacheFilter.QuickFilter`. A good usage example is method `MapSettingsUtils.showSettingsPopup()`.

The problematic thing of course is that you need to know for which context you want to access the filter settings. Because as you know we decided to store multiple settings depending on the context. Or, if the context is transient, you 

Wrt waypoints: also from `MapSettingsUtils.showSettingsPopup()` it seems to me that the setting used there is Settings.isExcludeWpVisited(). Strange that we have two settings here, maybe a duplicate?

**Comment by moving-bits** [comment]
Thanks @eddiemuc for providing those valuable pointers, will look into it. My idea is to let the popup (either cache or waypoint) query its calling activity (which would be either a `NewMap` or a `CGeoMap`) whether such a specific setting is active or not - the map knows their own current filter context and should be able to answer by either yes or no. Question now is: How can I call a method from the opening/calling activity?
By `getCallingActivity()` I can retrieve a component name, but sending an intent would create a new activity, wouldn't it?
Or can I somehow use the activity stack to call the calling activity?

Regarding the `getHideVisitedWaypoints` vs. `isExcludeWpVisited` I need to look into it. Maybe a duplicate, maybe I've simply chosen the wrong one in my tests...

**Comment by moving-bits** [comment]
Update: `getHideVisitedWaypoints` is for hiding the visited waypoints in the waypoints list of cache details. So I'd chosen the wrong method.
Choosing the other one I should be able to remove a visited waypoint from current route without calling the calling activity. Only updating the route display remains a problem in this case - as well as checking for "hide found caches" filter state.

**Comment by eddiemuc** [comment]
> My idea is to let the popup (either cache or waypoint) query its calling activity (which would be either a `NewMap` or a `CGeoMap`) whether such a specific setting is active or not

Hm I don't know how this would be possible. I see that CachePopupFragment has an getActivity method, but I looked into CachePopupFragment and saw that this is in turn shown by an "intermediate Activity" called CachePopup (which apparently does nothing else than showing the dialog). So I assume getActivity returns the CachePopup.

Sending a new intent would IMHO open a new activity, so this is not what you want.

How about simply passing the Filter as a parameter  in the intents extras by the calling Map Activity to the CachePopup(Fragment)? (`GeocacheFilter` can be serialized into a string using `toConfig()` and can be completely restored from it using `GeocacheFilter.createFromConfig()`) Or, if you're only interested in one or two specific settings, passing those parameters.

**Comment by moving-bits** [comment]
Thanks for cross-checking.

> How about simply passing the Filter as a parameter 

Yes, I guess, that'll be the way to go for the "hide waypoint" state.

For updating the route (which would need to call the map activity) I had another idea tonight: We have some broadcasting mechanism for cache updates, so that a sub-activity can inform other activities about a change of cache details. We could use a similar mechanism to let the map know that it should reload (and redraw) the individual route.

@Lineflyer 
As you had opened this issue, I have a few more detailed questions to you:
- Would it be ok to just check for setting the "visited" flag in the popup, or do we need to check the "visited" flag in the waypoint list of cache detail pages as well?
- If we go for the caches' found state as well: Which actions shall trigger this? AFAIK, changing a cache's found state can be done by using the menu ("set as found"), creating a specific offline log, and by sending a specific online log, not to mention the quick log feature - do we really want to check all of those places?


Another thing on my todo list is this: Currently I remove the waypoint from individual route as soon as someone checks the "is visited" flag (when the "hide visited waypoints" flag is set). I'll try to change this to remove the waypoint on closing the dialog only, so that if you hit the "is visited" flag accidentally, and remove it before closing the popup, the waypoint will not be removed from the individual route.

**Comment by ztNFny** [comment]
> * If we go for the caches' found state as well: Which actions shall trigger this? AFAIK, changing a cache's found state can be done by using the menu ("set as found"), creating a specific offline log, and by sending a specific online log, not to mention the quick log feature - do we really want to check all of those places?

It really is only 2 places, logOffline() and logVisit() in Geocache.

I tried implementing it there (to implement remove from route on (offline) log https://github.com/cgeo/cgeo/issues/13089) but as the UpdateIndividualRoute implementations are highly specific to the map implementation used that actually seems to require a big rewrite.

> Currently I remove the waypoint from individual route as soon as someone checks the "is visited" flag (when the "hide visited waypoints" flag is set).

I tried to find that as a base for removing it on log, but couldn't find anything in the codebase, I'd have expected this to be in toggleVisited.setOnClickListener ?

**Comment by moving-bits** [comment]
@ztNFny 
For notifying the different map views about updates to caches we use some broadcasting mechanism. Maybe we can use something similar to trigger a route update (so that the map views can take care of reloading and repainting the individual route for themselves)?

**Comment by ztNFny** [comment]
That sounds like a sensible approach, along with some static getRoute and isInRoute. Depending on the complexity it could also be implemented for the new map only.
